dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,254 @@
+
+import logging
+import inspect
+import math
+
+from typing import Callable, AnyStr, Any, List, Tuple, NamedTuple, Dict, Union, Optional
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from copy import deepcopy
+from time import perf_counter
+from collections import OrderedDict, namedtuple
+from enum import Enum
+
+import pandas as pd
+from more_itertools import chunked, flatten
+from tqdm.auto import tqdm as tqdm_auto
+
+from dkulib.io_utils.plugin_io_utils import generate_unique
+
+
+class ErrorHandling(Enum):
+    """Enum class to identify how to handle API errors"""
+
+    LOG = "Log"
+    FAIL = "Fail"
+
+
+class BatchError(ValueError):
+    """Custom exception raised if the Batch function fails"""
+
+
+class DataFrameParallelizer:
+    """Apply a function to a pandas DataFrame with parallelization, error logging and progress tracking.
+
+    This class is particularly well-suited for synchronous functions calling an API, either row-by-row or by batch.
+
+    Attributes:
+        function: Any function taking a dict as input (row-by-row mode) or a list of dict (batch mode),
+            and returning a response with additional information, typically a JSON string.
+            In batch mode, the response from the function should be parsable by the `batch_response_parser` attribute.
+        error_handling: If ErrorHandling.LOG (default), log the error from the function as a warning,
+            and add additional columns to the dataframe with the error message and error type.
+            If ErrorHandling.FAIL, the function will fail is there is any error.
+        exceptions_to_catch: Tuple of Exception classes to catch. Mandatory if ErrorHandling.LOG (default).
+        parallel_workers: Number of concurrent threads to parallelize the function. Default is 4.
+        batch_support: If True, send batches of row (list of dict) to the `function`
+            Else (default) send rows as dict to the function
+        batch_size: Number of rows to include in each batch. Default is 10.
+            Taken into account if `batch_support` is True.
+        batch_response_parser: Function used to parse the raw response from the function in batch mode,
+            and assign the actual responses and errors back to the original batch of row (list of dict).
+            This is often required for batch APIs which return nested objects with a mix of responses and errors.
+        output_column_prefix: Column prefix to add to the output columns of the dataframe,
+            containing the `function` responses and errors. Default is "output".
+        verbose: If True, log additional information on errors.
+            Else (default) log the error message and the error type.
+
+    """
+
+    DEFAULT_PARALLEL_WORKERS = 4
+    DEFAULT_BATCH_SIZE = 10
+    DEFAULT_BATCH_SUPPORT = False
+    DEFAULT_VERBOSE = False
+    DEFAULT_OUTPUT_COLUMN_PREFIX = "output"
+    OUTPUT_COLUMN_NAME_DESCRIPTIONS = OrderedDict(
```

Comment:
I sure can move them to another file. Are you convinced it would make the code more readable? 
I kinda enjoy having constants at class level, so that I remember they are linked to a specific class. Maybe that's a bad habit?

Regarding removing them as "arguments from the constructor" what do you mean?

Reply:
I can change, I just wasn't aware of this good practice.

Just to understand, what do you mean by "reduce the number of arguments in the constructor"?

Reply:
I think you could delete the following argument of the constructor and corresponding attributes : `parallel_workers`, `batch_support` `batch_size` `output_column_prefix` and `output_column_prefix`

Reply:
Do you mean that these arguments should be made into constants?

I am asking because
- `parallel_workers`, `batch_support`, `batch_size` can be tuned by the plugin preset admin
- `output_column_prefix` is usually changed by the plugin developer for each API endpoint

Reply:
True... Then it makes no sense to transform the params to constants. Let's forget this idea

Reply:
Oki! It's true that these are not very clear in the code, I'll add comments

Reply:
Docstrings added in 21adc15